File: VeilLink/Security/Base32.swift
import Foundation

enum Base32 {
    private static let alphabet = Array("ABCDEFGHIJKLMNOPQRSTUVWXYZ234567")

    static func encode(_ data: Data) -> String {
        var output = ""
        var buffer = 0
        var bitsLeft = 0

        for byte in data {
            buffer = (buffer << 8) | Int(byte)
            bitsLeft += 8
            while bitsLeft >= 5 {
                bitsLeft -= 5
                output.append(alphabet[(buffer >> bitsLeft) & 31])
            }
            if bitsLeft == 0 {
                buffer = 0
            } else {
                buffer &= (1 << bitsLeft) - 1
            }
        }

        if bitsLeft > 0 {
            output.append(alphabet[(buffer << (5 - bitsLeft)) & 31])
        }
        return output
    }
}

File: android/protocol-core/src/main/kotlin/studio/zeo/veillink/protocol/Base32.kt
package studio.zeo.veillink.protocol

object Base32 {
    private const val ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567"

    fun encode(data: ByteArray): String {
        val out = StringBuilder()
        var buffer = 0
        var bitsLeft = 0
        for (raw in data) {
            buffer = (buffer shl 8) or (raw.toInt() and 0xff)
            bitsLeft += 8
            while (bitsLeft >= 5) {
                bitsLeft -= 5
                out.append(ALPHABET[(buffer shr bitsLeft) and 31])
            }
            buffer = if (bitsLeft == 0) 0 else buffer and ((1 shl bitsLeft) - 1)
        }
        if (bitsLeft > 0) out.append(ALPHABET[(buffer shl (5 - bitsLeft)) and 31])
        return out.toString()
    }
}
